rfc/state-model.rst

```diff
@@ -0,0 +1,299 @@
+
+For a firmware component, depending on the state or progress of a firmware update, there can be one or more firmware images currently in the firmware store:
+
+1. An image that is actively in use by the system.
+2. An image that is being prepared for installation.
+3. A 'backup' of a previously installed image, used to recover if an attempted update fails.
+4. An expired, rejected, or incomplete image, that can be erased.
+
+Assuming that the firmware component is essential for system operation, there will always be exactly one image of type (1). Other images might, or might not, be present in the store.
+
+Although a Firmware store in a specific system might have storage for more than two firmware images, we define a state model for the Firmware Store that only requires two. This is possible because the store does not need to hold more than one firmware image of type (2), (3), or (4) concurrently.
+
+Instead of identifying a physical storage location (sometimes referred to as a 'bank' or 'slot') for the firmware images, the API uses the following (*working names*) for the two required locations:
+
+========  =========  ===========
+Location  Present    Description
+========  =========  ===========
+*Active*  Always     The image that is actively in use by the system
+*Second*  Sometimes  An image that is being prepared, or is kept for recovery, or needs to be erased
+========  =========  ===========
+
+Depending on the system and memory design, the *active* and *second* locations can be fixed physical storage locations, or can refer to different physical storage locations over time as an update progresses. The implementation of the API is responsible for mapping the logical storage locations to the stored firmware images.
+
+During the course of an update, a specific firmware image can change from being *active* to *second*, or from *second* to *active*. For example:
+
+* An image will switch from being *second* --- while being prepared --- to *active* following installation.
+* An image will switch from being *active* to *second* when it becomes the backup image during installation of new firmware.
+
+Firmware Store state model
+--------------------------
+
+A persistent state model for an in-progress Firmware Update is required for two reasons:
+
+1. When updating firmware that is essential for system operation, a system restart is required to complete installation, and start executing the new system firmware. Communication between the runtime system and the Bootloader regarding the required installation operations must be done via memory.
+2. On constrained devices, the download of a new firmware image may take a long time, due to bandwidth or power limitations. A device restart should not invalidate the update progress that has already occurred.
```

Comment:
Does this mean that if an image download is not finished before a restart, the next time bootup, the image update process should start from where it is broken before the reboot?

Reply:
This was the intention in the existing specification, and there are good reasons to retain this use case for v1.0. This makes the API usable for devices where there is limited bandwidth or energy for downloaded updates:
* The image can be transferred over multiple restarts of the device and stored bit by bit
* Once the image is fully downloaded, it can be installed

The API does not provide any support for querying what data has been written to an image. Specifying such an API is difficult without defining how the Client must write the data (e.g. streaming, or arbitrarily-ordered aligned blocks, etc.). Allowing flexibility in how the data is written was an important goal in v0.7.

The implication is that for a Client to be able to restart/continue writing data to an image, it (or the Update Server) must keep track of what data has already been successfully written.

The FWU API also supports implementations where a partially written image is lost at reboot. For a component like this, `reboot` would transition a WRITING state to either FAILED (needs erase) or READY, depending on whether the image is initially written to non-volatile memory or to RAM.

Reply:
As much as I think it's a good idea to be able to continue the update after an reboot it may be difficult to implement in a safe way. The write function, as it is now, doesn't give any guarantees of what's happening with the data. They may be buffered or written to ram which can be erased after a reboot (or a crash). To make it safe we would need to put some guarantees around write behaviour, but this may lead to complications where they are not needed.

Reply:
@bulislaw - Perhaps, but there is clearly intention in the specification to enable an implementation that preserves partial images across a reboot.

To make it possible for a Client to reason about this behavior, we might need to put some conditions on what is guaranteed by the API. For example, we could require that:
* For an implementation that has a persistent _second_ image, when `psa_fwu_write()` returns `PSA_SUCCESS`, then the data is persisted (i.e. is present following a subsequent reboot)
* If the system restarts before `psa_fwu_write()` returns `PSA_SUCCESS`, or the call returns an error, then it is unknown whether, or how much of, the data was persisted.

I wouldn't want to make `psa_fwu_write()` atomic: it is impossible to insist on flash memory being returned to the state it was in prior to a write which did not completely success (even if we keep a copy of all the data that is overwritten). And using a journalling file system for this does not make sense.

However, depending on the storage memory used for the _second_ image, it might be OK for the Client to write the same data to the same location in the image twice. I.e., following a failed write (e.g. due to crash), the Client can attempt the same write again. If this is not possible, then a failed/rebooted `write` operation should result in the Client calling `erase` and starting again.

Reply:
If the secondary slot is persistent, is starting from where it is broken before the reboot a requirement or an optional choice?
My concern is that to achieve the break and then recovery, the OTA server is required to cooperate.  Which data block comes from the server is not decided by the OTA client. Take the AWS OTA server for example, if data block of the image is not finished before a reboot, then the next time reboots, after OTA client tells the server that an OTA process can be started now(via MQTT), then the server starts the image download again just like that the file has never been downloaded. Also, in some data transport protocols like MQTT, the blocks of the image may come to the client out of order. It hardens FWU service to record which block has been downloaded. Without the cooperation of OTA server, partial download cannot be recovered after a reboot.
Better explain the responsibility/requirement of OTA server and the FWU service and OTA agent to achieve the download recovery.

Reply:
> Better explain the responsibility/requirement of OTA server and the FWU service and OTA agent to achieve the download recovery.

@Sherryzhang2 - I think that is a good idea. For all of the interesting use cases that have driven some non-trivial aspect of the API design, it would be valuable to describe the use case, illustrate how the Client can use the API to achieve this use case, and list the requirements on the OTA Server, Client, and Service for the use case.

The OTA Server and/or client (depending on which entity is driving the update process) has to be able to cooperate for download resumption after a restart. If the OTA protocol does not support resuming an update session, then the Client can `erase` the partial update, and start again.